### PR TITLE
Feat/smart router

### PR DESCRIPTION
**Pull request by jkrivine**
https://excalidraw.com/#room=5328a7cb626cf8976a1c,MBClLetW89XuLJJFR0SQQA

### CODE REVIEW

File: src/strategies/routing_logic/ERC4626Logic.sol.wip [hunk 1]
```diff
@@ -0,0 +1,29 @@
+// SPDX-License-Identifier:	BSD-2-Clause
```
Comment (lnist) [inline src/strategies/routing_logic/ERC4626Logic.sol.wip:1]: I know this is a draft commit, but I think we should leave these .wip files on a separate branch. I have not looked at them.

File: src/strategies/vendor/IERC4626.sol [hunk 2]
```diff
@@ -0,0 +1,227 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.20;
+
+import {IERC20} from "@mgv/lib/IERC20.sol";
+
+/**
+ * @dev Interface of the ERC4626 "Tokenized Vault Standard", as defined in
+ * https://eips.ethereum.org/EIPS/eip-4626[ERC-4626].
+ */
+interface IERC4626 is IERC20 {
```
Comment (lnist) [inline src/strategies/vendor/IERC4626.sol:10]: If this is unused, then I think it should be a separate PR. Same for stargate vendor files

File: src/strategies/routers/SmartRouter.sol [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+  ///@param logic the loggic to use for the given offer.
+  function setLogic(IERC20 token, OfferCoordinates calldata oc, AbstractRouter logic) external onlyAdmin {
+    SmartRouterStorage.getStorage().routeLogics[token][oc.olKeyHash][oc.offerId] = logic;
+    emit SetRouteLogic(token, oc.olKeyHash, oc.offerId, logic);
+  }
+
+  /// @inheritdoc AbstractRouter
+  /// @dev pulls liquidity using the logic specified by offer owner
+  /// @dev data must encode OfferCoordinates
+  function __pull__(IERC20 token, uint amount, bytes memory data) internal virtual override returns (uint) {
```
Comment (lnist) [inline src/strategies/routers/SmartRouter.sol:33]: If the only usage of the abi-encoded data is to let SmartRouter have the same API as router logics then I think they should instead have separate interfaces. They are not pluggable the same places anyways since the caller has different parameters.

File: src/strategies/routers/SmartRouterStorage.sol [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+import {IERC20} from "@mgv/lib/IERC20.sol";
+import {AbstractRouter} from "./abstract/AbstractRouter.sol";
+
+///@title Mangrove Smart Router storage (randomized access)
+library SmartRouterStorage {
+  bytes32 private constant OFFSET = keccak256("MangroveStrats.SmartRouterStorage.Layout");
+  ///@notice Storage layout
+
+  ///@notice routeLogics logics approved by admin in order to pull/push liquidity in an offer specific manner
```
Comment (lnist) [inline src/strategies/routers/SmartRouterStorage.sol:12]: We seem to be calling them AbstractRouters, routeLogics, and logics (which can be confused with offer logics) - we should stick to a single term.

File: src/strategies/routing_logic/AaveLogic.sol [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+import {SmartRouter, AbstractRouter} from "@mgv-strats/src/strategies/routers/SmartRouter.sol";
+import {IERC20} from "@mgv/lib/IERC20.sol";
+import {AaveMemoizer} from "@mgv-strats/src/strategies/integrations/AaveMemoizer.sol";
+import {AaveLogicStorage} from "./AaveLogicStorage.sol";
+import {TransferLib} from "@mgv/lib/TransferLib.sol";
+
+/// @title AaveLogic
+/// @notice Routing logic for Aave
+contract AaveLogic is AbstractRouter, AaveMemoizer {
```
Comment (lnist) [inline src/strategies/routing_logic/AaveLogic.sol:12]: I have not looked at this yet. Will look when other stuff is resolved :)

File: src/strategies/MangroveOrder.sol [hunk 6]
```diff
@@ -122,7 +122,7 @@ contract MangroveOrder is Forwarder, IOrderLogic {
     // Pulling funds from `msg.sender`'s reserve
     // `takerGives` is derived via same function as in `execute` of core protocol to ensure same behavior.
     uint takerGives = tko.fillWants ? tko.tick.inboundFromOutboundUp(tko.fillVolume) : tko.fillVolume;
-    uint pulled = router().pull(IERC20(tko.olKey.inbound_tkn), msg.sender, takerGives, true);
+    uint pulled = router().pull(IERC20(tko.olKey.inbound_tkn), takerGives, abi.encode(true, msg.sender));
```
[the remark this answers is not in the document]
Comment (jkrivine) [reply src/strategies/MangroveOrder.sol:125]: Right now, MangroveOrder is still working with a SimpleRouter. When we change this, you are right

File: src/strategies/integrations/AaveMemoizer.sol [hunk 7]
```diff
@@ -0,0 +1,229 @@
+// SPDX-License-Identifier:	BSD-2-Clause
```
[the remark this answers is not in the document]
Comment (jkrivine) [reply src/strategies/integrations/AaveMemoizer.sol:1]: yes

File: src/strategies/routers/SmartRouter.sol [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+    }
+    return amount;
+  }
+
+  ///@inheritdoc AbstractRouter
+  ///@dev data must encode OfferCoordinates
+  function balanceOfReserve(IERC20 token, bytes calldata data) public view override returns (uint) {
+    OfferCoordinates memory oc = abi.decode(data, (OfferCoordinates));
+    AbstractRouter logic = SmartRouterStorage.getStorage().routeLogics[token][oc.olKeyHash][oc.offerId];
+    return logic.balanceOfReserve(token, bytes(""));
```
Comment (maxencerb) [inline src/strategies/routers/SmartRouter.sol:75]: Here, we may want to have a fallback behavior if this is address 0 (i.e. token.balanceOf(user))

File: src/strategies/routers/SmartRouter.sol [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+    return logic.checkList(token, bytes(""));
+  }
+
+  ///@inheritdoc AbstractRouter
+  ///@dev data must encode OfferCoordinates
+  function __activate__(IERC20 token, bytes calldata data) internal override {
+    OfferCoordinates memory oc = abi.decode(data, (OfferCoordinates));
+    AbstractRouter logic = SmartRouterStorage.getStorage().routeLogics[token][oc.olKeyHash][oc.offerId];
+    return logic.activate(token, bytes(""));
+  }
```
Comment (maxencerb) [inline src/strategies/routers/SmartRouter.sol:93]: Here we could check if there is router to avoid a call (2600 gas) even though this implementation works

File: src/strategies/MangroveOrder.sol [hunk 10]
```diff
@@ -119,11 +106,19 @@ contract MangroveOrder is Forwarder, IOrderLogic {
...
+      amount: tko.fillWants ? tko.tick.inboundFromOutboundUp(tko.fillVolume) : tko.fillVolume,
+      fundOwner: msg.sender,
+      token: IERC20(tko.olKey.inbound_tkn)
+    });
+    (RouterProxy proxy,) = ROUTER_FACTORY.instantiate(msg.sender, ROUTER_IMPLEMENTATION);
+    SmartRouter userRouter = SmartRouter(address(proxy));
+    if (tko.takerGivesLogic != AbstractRouter(address(0))) {
+      userRouter.setLogic(pullOrder, tko.takerGivesLogic);
+    }
+    require(userRouter.pull(pullOrder, true) == pullOrder.amount, "mgvOrder/transferInFail");
```
Comment (maxencerb) [inline src/strategies/MangroveOrder.sol:121]: maybe the `instantiate` function should return an already casted instance of `SmartRouter` so it's clear what we get. We have no interaction with the proxy anyway. Something like that:

```suggestion
    (SmartRouter userRouter,) = ROUTER_FACTORY.instantiate(msg.sender, ROUTER_IMPLEMENTATION);
    if (tko.takerGivesLogic != AbstractRouter(address(0))) {
      userRouter.setLogic(pullOrder, tko.takerGivesLogic);
    }
    require(userRouter.pull(pullOrder, true) == pullOrder.amount, "mgvOrder/transferInFail");
```

File: src/strategies/offer_forwarder/abstract/ExpirableForwarder.sol [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+  ///@param olKeyHash the hash of the offer list key. It is indexed so RPC call can filter on it.
+  ///@param offerId the Mangrove offer id.
+  ///@param date in seconds since unix epoch
+  ///@notice By emitting this data, an indexer will be able to keep track of the expiry date of an offer.
+  event SetExpiry(bytes32 indexed olKeyHash, uint indexed offerId, uint date);
+
+  ///@notice `_expiryMaps[olKey.hash()][offerId]` gives timestamp beyond which `offerId` on the `olKey.(outbound_tkn, inbound_tkn, tickSpacing)` offer list should renege on trade.
+  ///@notice if the order tx is included after the expiry date, it reverts.
+  ///@dev 0 means no expiry.
+  mapping(bytes32 olKeyHash => mapping(uint offerId => uint expiry)) private _expiryMaps;
```
Comment (maxencerb) [inline src/strategies/offer_forwarder/abstract/ExpirableForwarder.sol:26]: No way to pack this in `OwnerData` ?

File: src/strategies/offer_forwarder/abstract/Forwarder.sol [hunk 12]
```diff
@@ -113,6 +116,11 @@ abstract contract Forwarder is IForwarder, MangroveOffer {
     require(owner != address(0), "Forwarder/unknownOffer");
   }
 
+  /// @inheritdoc IOfferLogic
+  function router(address fundOwner) public view override returns (AbstractRouter) {
+    return AbstractRouter(address(ROUTER_FACTORY.computeProxyAddress(fundOwner, ROUTER_IMPLEMENTATION)));
```
Comment (maxencerb) [inline src/strategies/offer_forwarder/abstract/Forwarder.sol:121]: Here same comment as above, why shouldn't we return an `AbstracRouter` instance ?

File: src/strategies/routers/RouterProxyFactory.sol [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+
+/// @title Mangrove Router Proxy Factory
+/// @notice Factory contract for the deployment of RouterProxy instances using CREATE2 for deterministic addresses.
+/// @dev Utilizes Ethereum's CREATE2 opcode for deploying contracts with predictable addresses.
+contract RouterProxyFactory {
+  /// @notice Emitted when a new proxy is deployed through this factory.
+  /// @param proxy the deployed proxy contract
+  /// @param owner The address which will be the admin and immutable owner of the newly deployed proxy.
+  /// @param implementation The address of the router implementation used by the proxy.
+  event ProxyDeployed(RouterProxy proxy, address indexed owner, AbstractRouter indexed implementation);
```
Comment (maxencerb) [inline src/strategies/routers/RouterProxyFactory.sol:15]: I would prefer something like this for indexing :

```solidity
event ProxyDeployed(address indexed owner, RouterProxy proxy, AbstractRouter implementation);
```

The only indexed field needed would be the the owner as this event is grabbed directly from the router contract. The rest can be data I think

File: src/strategies/routers/RouterProxyFactory.sol [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+  /// @dev The computed address is determined by the owner's address and the factory's address
+  function computeProxyAddress(address owner, AbstractRouter routerImplementation)
+    public
+    view
+    returns (address payable)
+  {
+    bytes memory creationCode = type(RouterProxy).creationCode;
+    bytes memory args = abi.encode(routerImplementation);
+    bytes32 initcodeHash = keccak256(abi.encodePacked(creationCode, args));
+    bytes32 salt = keccak256(abi.encode(owner));
```
Comment (maxencerb) [inline src/strategies/routers/RouterProxyFactory.sol:30]: maybe internal function getter something like (code repetition)

```solidity
function _getSalt(address owner) private pure returns (bytes32 salt) {
  salt = keccak256(abi.encode(owner)); 
}
```

File: src/strategies/routers/RouterProxyFactory.sol [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+  }
+
+  /// @notice Deploys a new RouterProxy for a given owner.
+  /// @param owner The address to be set as initial admin and immutable owner of the proxy.
+  /// @param routerImplementation router contract which implements routing functions
+  /// @return proxy The address of the newly deployed RouterProxy.
+  /// @dev Emits a ProxyDeployed event upon successful deployment.
+  ///      Note that the deployment can be initiated by any caller, on behalf of `owner`.
+  function deployProxy(address owner, AbstractRouter routerImplementation) public returns (RouterProxy proxy) {
+    proxy = new RouterProxy{salt:keccak256(abi.encode(owner))}(routerImplementation);
```
Comment (maxencerb) [inline src/strategies/routers/RouterProxyFactory.sol:48]: use the newly created `_getSalt(owner)` function as well

File: src/strategies/routers/SmartRouter.sol [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+import {SimpleRouter, RL} from "./SimpleRouter.sol";
+
+///@title Mangrove Smart Router implementation
+contract SmartRouter is SimpleRouter {
+  ///@notice logs new setting of routing logic for a specific offer
+  ///@param token the asset whose route is being set
+  ///@param olKeyHash the hash of the offer list
+  ///@param offerId the offer identifier for which the routing is set
+  ///@param logic the contract implementing the pull and push function for that specific route
+  event SetRouteLogic(IERC20 indexed token, bytes32 indexed olKeyHash, uint offerId, AbstractRouter logic);
```
Comment (maxencerb) [inline src/strategies/routers/SmartRouter.sol:15]: Maybe we should index `offerId` as well ?

```solidity
event SetRouteLogic(IERC20 indexed token, bytes32 indexed olKeyHash, uint indexed offerId, AbstractRouter logic);
```

File: src/strategies/routers/SmartRouter.sol [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+    AbstractRouter logic =
+      SmartRouterStorage.getStorage().routeLogics[routingOrder.token][routingOrder.olKeyHash][routingOrder.offerId];
+    if (address(logic) != address(0)) {
+      // delegating routing to the specific chosen route
+      (bool success, bytes memory retdata) =
+        address(logic).delegatecall(abi.encodeWithSelector(AbstractRouter.pull.selector, routingOrder, strict));
+      if (!success) {
+        SmartRouterStorage.revertWithData(retdata);
+      }
+      return (abi.decode(retdata, (uint)));
```
Comment (maxencerb) [inline src/strategies/routers/SmartRouter.sol:40]: We can maybe check the low-level way to delegate call to save some gas. Even if it's low level, it's a very well known design, we could simply have a `_delegate()` function to use in both pull and push like in the openzeppelin library `Proxy` implementation.

File: src/strategies/routers/SmartRouter.sol [hunk 18]
```diff
@@ -0,0 +1,83 @@
...
+  /// @dev pushes liquidity using the logic specified by offer owner
+  function __push__(RL.RoutingOrder memory routingOrder) internal virtual override returns (uint) {
+    AbstractRouter logic =
+      SmartRouterStorage.getStorage().routeLogics[routingOrder.token][routingOrder.olKeyHash][routingOrder.offerId];
+    if (address(logic) != address(0)) {
+      (bool success, bytes memory retdata) =
+        address(logic).delegatecall(abi.encodeWithSelector(AbstractRouter.push.selector, routingOrder));
+      if (!success) {
+        SmartRouterStorage.revertWithData(retdata);
+      }
```
Comment (maxencerb) [inline src/strategies/routers/SmartRouter.sol:57]: Same as above

File: src/strategies/routers/SmartRouter.sol [hunk 19]
```diff
@@ -0,0 +1,83 @@
...
+    if (address(logic) != address(0)) {
+      (bool success, bytes memory retdata) = address(this).staticcall(
+        abi.encodeWithSelector(
+          SmartRouterStorage._staticdelegatecall.selector,
+          abi.encodeWithSelector(AbstractRouter.balanceOfReserve.selector, routingOrder)
+        )
+      );
+      if (!success) {
+        SmartRouterStorage.revertWithData(retdata);
+      }
```
Comment (maxencerb) [inline src/strategies/routers/SmartRouter.sol:77]: same here

File: src/strategies/routers/RouterProxy.sol [hunk 20]
```diff
@@ -0,0 +1,48 @@
...
+
+  /// @notice Deploys a Proxy for the SmartRouter that handles incoming transactions and delegates them to the implementation.
+  /// @param implementation The address of the deployed SmartRouter contract to which calls will be delegated.
+  /// @dev Initializes the contract with an AccessControlled base to set up access control.
+  constructor(AbstractRouter implementation) {
+    IMPLEMENTATION = implementation;
+    // store the msg sender at address 0 (_admin storage slot on access controlled)
+    assembly {
+      sstore(0, caller())
+    }
```
Comment (maxencerb) [inline src/strategies/routers/RouterProxy.sol:23]: Missing SetAdmin(address admin) event throwing

**Review by espendk — APPROVED** [review]
This PR has been extensively reviewed by @jkrivine and @lnist + it's been deemed ready for audit, so I agree that it's ready to merge to `develop`.
